### PR TITLE
#161348800 Like/Dislike article

### PR DESCRIPTION
**Pull request by judeinno**
#### What does this PR do?
- Ensure that users can like/dislike an article.
- Ensure that the user can update like status.
- Refactor test to use basetest file.

#### Description of Task to be completed?
Have the following endpoints working;
`POST|GET /api/articles/<int:article_id>/like_status`
`PUT /api/articles/<int:article_id>/like_status`

#### How should this be manually tested?
Using post man to test the endpoints above
To like an article
URL: `POST|GET /api/articles/<int:article_id>/like_status`
Payload
````
{
	"like_status": "dislike"
}
````
Response:
````
{
    "Like_status": {
        "id": 1,
        "article_title": "How to train your dragon",
        "like_status": "like",
        "article": 1,
        "user":  "judeinno"
    }
}
````
To update the like status
URL: `PUT /api/articles/<int:article_id>/like_status`
Payload
````
{
	"like_status": "like"
}
````
Response
````
{
    "Like_status": {
        "like_status": "dislike"
    }
}
````
#### What are the relevant pivotal tracker stories?
[#161348800](https://www.pivotaltracker.com/story/show/161348800)

### CODE REVIEW

**Review by emugaya — CHANGES_REQUESTED** [review]
@judeinno Have you added the like statistics to the article?  A user should be able to see how many likes and dislikes they have on their article and that information should be  returned together with the article.

File: authors/apps/articles/models.py [hunk 1]
```diff
@@ -19,3 +19,19 @@ class Article(models.Model):
 
     def __str__(self):
         return self.title
+
+
+class LikeArticle(models.Model):
+    LIKE = 'like'
```
Comment (emugaya) [inline authors/apps/articles/models.py:43]: What are you trying to achieve here???

File: authors/apps/articles/renderers.py [hunk 2]
```diff
@@ -26,3 +26,21 @@ def render(self, data, media_type=None, renderer_context=None):
                 'articles': data,
                 'articlesCount': len(Article.objects.all())
             })
+
+
+class LikeStatusRenderer(JSONRenderer):
+    charset = 'utf-8'
+
+    def render(self, data, media_type=None, renderer_context=None):
```
Comment (emugaya) [inline authors/apps/articles/renderers.py]: What does this method do?

File: authors/apps/articles/tests_articles/test_articles.py [hunk 3]
```diff
@@ -150,3 +92,74 @@ def test_delete_article(self):
...
+        self.test_create_article()
+        article = Article.objects.all().first()
+        like_status_url = "/api/articles/{}/like_status".format(article.pk)
+        self.client.post(like_status_url,
+                         self.like_status_data,
+                         format='json')
+        response = self.client.get(like_status_url, format='json')
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_get_like_status_article_doesnt_exist(self):
```
Comment (emugaya) [inline authors/apps/articles/tests_articles/test_articles.py]: Assert for status code here

File: authors/apps/articles/tests_articles/test_articles.py [hunk 4]
```diff
@@ -150,3 +92,74 @@ def test_delete_article(self):
...
+                                    self.like_status_data,
+                                    format='json')
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+
+    def test_like_status_article_doesent_exist(self):
+        like_status_url = "/api/articles/{}/like_status".format(100)
+        response = self.client.post(like_status_url,
+                                    self.like_status_data,
+                                    format='json')
+        self.assertIn("Article Not found",
```
Comment (emugaya) [inline authors/apps/articles/tests_articles/test_articles.py]: Assert for status code

File: authors/apps/articles/views.py [hunk 5]
```diff
@@ -119,3 +124,78 @@ class ArticleRetrieveBySlugAPIView(generics.RetrieveAPIView):
...
+    post: Create an  like or dislike.
+    """
+    serializer_class = LikeSerializer
+    permission_classes = (permissions.IsAuthenticatedOrReadOnly,)
+    renderer_classes = (LikeStatusRenderer,)
+    lookup_field = 'pk'
+
+    def post(self, request, *args, **kwargs):
+        if not Article.objects.filter(pk=kwargs['pk']).exists():
+            raise NotFound(detail="Article Not found", code=404)
```
Comment (emugaya) [inline authors/apps/articles/views.py]: No need to pass `code=404`. This is already provided for you by the exception

File: authors/apps/articles/views.py [hunk 6]
```diff
@@ -119,3 +124,78 @@ class ArticleRetrieveBySlugAPIView(generics.RetrieveAPIView):
...
+        return Response(serializer.data, status=status.HTTP_201_CREATED)
+
+    def get(self, request, *args, **kwargs):
+        if not Article.objects.filter(pk=kwargs['pk']).exists():
+            raise NotFound(detail="Article Not found", code=404)
+        if LikeArticle.objects.filter(article=kwargs['pk']).exists():
+            like_status = LikeArticle.objects.filter(article=kwargs['pk'])
+            like_statuses = LikeSerializer(like_status, many=True)
+            return Response(like_statuses.data, status=status.HTTP_200_OK)
+        return Response(
```
Comment (emugaya) [inline authors/apps/articles/views.py]: Raise a Bad Request here.

File: authors/apps/articles/views.py [hunk 7]
```diff
@@ -119,3 +124,78 @@ class ArticleRetrieveBySlugAPIView(generics.RetrieveAPIView):
...
+class UpdateLikeStatusAPIView(generics.RetrieveUpdateAPIView):
+    serializer_class = LikeSerializer
+    permission_classes = (permissions.IsAuthenticatedOrReadOnly,)
+    renderer_classes = (LikeStatusRenderer,)
+    lookup_field = 'pk'
+
+    def put(self, request, *args, **kwargs):
+        article_id = kwargs['pk']
+        if not Article.objects.filter(pk=article_id).exists():
+            raise NotFound(detail="Article Not found", code=404)
```
Comment (emugaya) [inline authors/apps/articles/views.py]: No need for this `code=404` Read about exceptions.

File: authors/apps/articles/models.py (hunk 1 above)
Comment (judeinno) [reply authors/apps/articles/models.py:43]: I was creating the choices to be passed, So its like true or false but i then used like or dislike

File: authors/apps/articles/renderers.py (hunk 2 above)
Comment (judeinno) [reply authors/apps/articles/renderers.py]: When come one likes an article its the renderer that renders my data, It returns the Like_status_id, the article title and id plus the username of the person choosing to like/dislike an article

**Review by brucemakallan — COMMENTED** [review]
For the Hound violations of "Line too long", you may use this `\` to split the line. e.g:
```
 if LikeArticle.objects.filter(user=user) \
             .filter(article=article_id).exists():
```

File: authors/apps/articles/tests_articles/test_articles.py (hunk 3 above)
Comment (judeinno) [reply authors/apps/articles/tests_articles/test_articles.py]: Done

File: authors/apps/articles/tests_articles/test_articles.py (hunk 4 above)
Comment (judeinno) [reply authors/apps/articles/tests_articles/test_articles.py]: Done

File: authors/apps/articles/views.py (hunk 5 above)
Comment (judeinno) [reply authors/apps/articles/views.py]: Done

File: authors/apps/articles/views.py (hunk 6 above)
Comment (judeinno) [reply authors/apps/articles/views.py]: Working on it, django rest framework site currently down

File: authors/apps/articles/views.py (hunk 7 above)
Comment (judeinno) [reply authors/apps/articles/views.py]: Done

**Comment by judeinno** [comment]
@emugaya The user can now see the number of likes and dislikes when they get an article

**Comment by judeinno** [comment]
@brucemakallan I have also fixed the hound violations

**Comment by brucemakallan** [comment]
> @brucemakallan I have also fixed the hound violations

Thanks

File: authors/apps/articles/views.py (hunk 6 above)
Comment (judeinno) [reply authors/apps/articles/views.py]: Done working on it

**Pull request merged**